### PR TITLE
feat: return better error if file size is too big to upload

### LINKED ISSUE #7071: Show better error when template is too big
```
$ truncate -s 1G bigbig
$ coder templates push -v
> Upload "."? (yes/no)
Failed to read file from request.
```

This isn't very helpful.

### PR DESCRIPTION
**Pull request by Emyrk**
fixes: https://github.com/coder/coder/issues/7071

The issue was the error condition was only checked after the bytes have been written. So the coderd side was returning an api error for the content being too large before the cli could do the same client side check.

To fix this, I just check if we are going to exceed the limit with the `fileInfo.Size` before we write.

### CODE REVIEW

**Review by mafredri — COMMENTED** [review]
Does this actually fix the underlying problem? AFAICT we're not measuring actual tar size? So there's a small range of inputs where total file size < limit but add tar overhead (prefix, file names) and it exceeds the limit? Or does `Write` take that into account?

File: provisionersdk/archive.go [hunk 1]
```diff
@@ -106,7 +111,7 @@ func Tar(w io.Writer, directory string, limit int64) error {
 		}
 		totalSize += wrote
 		if limit != 0 && totalSize >= limit {
```
[the remark this answers is not in the document]
Comment (mafredri) [reply provisionersdk/archive.go:113]: Perhaps we can wrap `w` (`wc := writeCounter{w}`) before passing it on to tar writer, and count bytes written to it? Then we don't need to track `wrote` here and just check `wc.written`.

[the remark this answers is not in the document]
Comment (Emyrk) [reply provisionersdk/archive.go:113]: You are right about the headers. Honestly I didn't look too deep at the code and just assumed the `totalSize` was being tracked correctly :facepalm:.

As for the `<=` vs `>=`, I'll change the conditional to `>` on the check to be consistent with the api.

[the remark this answers is not in the document]
Comment (Emyrk) [reply provisionersdk/archive.go]: @mafredri good catch. I am using a limit writer now

[the remark this answers is not in the document]
Comment (Emyrk) [reply provisionersdk/archive.go]: Writing a test or two

**Review by mafredri — APPROVED** [review]
Nice improvement 👍

**Pull request merged**